### PR TITLE
Feature marc newoutputs shortageconstraint multicriteria

### PR DESCRIPTION
**Pull request by marc-juanpera**
I didn't succeed, in reasonable time, to separate all contributions on different branches. 
I will prepare myself to explain them quickly when asked, and I will do a pull request for each concrete change from now on.

### CODE REVIEW

**Comment by smartie2076** [comment]
I now updated the Offgridders file structure with pull request #36, which will now require merging your feature with the new structure. Also added some tests to make sure the code doesn't break.

File: A0_main_script.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
     input_excel_file = str(sys.argv[1])
 else:
     # generic input file
-    input_excel_file = './inputs/test_input_template.xlsx'
+    input_excel_file = './inputs/input_for_multicriteria_peru.xlsx'
```
Comment (smartie2076) [inline A0_main_script.py]: Does that mean the input file changes completely? 
Then please present me this xlsx in out meeting next week.

File: C_sensitivity_experiments.py [hunk 2]
```diff
@@ -523,6 +525,8 @@ def overall_results_title(settings, number_of_project_sites, sensitivity_array_d
             'lcoe',
             'annuity',
             'npv',
+            'first_investment',
+            'operation_mantainance_expenditures',
```
Comment (smartie2076) [inline C_sensitivity_experiments.py]: OK

File: C_sensitivity_experiments.py [hunk 3]
```diff
@@ -555,10 +559,13 @@ def overall_results_title(settings, number_of_project_sites, sensitivity_array_d
...
             'power_storage_kW',
             'capacity_rectifier_ac_dc_kW',
-            'capacity_inverter_kW',
+            'capacity_inverter_dc_ac_kW',
             'capacity_wind_kW',
             'capacity_genset_kW',
             'capacity_pcoupling_kW',
+            'total_pv_generation_kWh',
+            'total_wind_generation_kWh',
+            'total_genset_generation_kWh',
```
Comment (smartie2076) [inline C_sensitivity_experiments.py]: 562, 566, 567, 568 OK

File: G2a_oemof_busses_and_componets.py [hunk 4]
```diff
@@ -320,9 +320,9 @@ def storage_fix(micro_grid_system, bus_electricity_dc, experiment, capacity_stor
             outputs={bus_electricity_dc: solph.Flow(
                 nominal_value = power_storage #capacity_storage*experiment['storage_Crate_discharge']
                 )},  # maximum discharge possible in one timestep
-            capacity_loss               = experiment['storage_loss_timestep'],  # from timestep to timestep
-            capacity_min                = experiment['storage_soc_min'],
-            capacity_max                = experiment['storage_soc_max'],
+            loss_rate               = experiment['storage_loss_timestep'],  # from timestep to timestep
+            min_storage_level                = experiment['storage_soc_min'],
+            max_storage_level                = experiment['storage_soc_max'],
```
Comment (smartie2076) [inline G2a_oemof_busses_and_componets.py]: Are you using oemof 0.3.0 on your machine and the Offgridders environment? The current used version is oemof 0.2.x, and these parameters (loss rate, min_storage_level, max_storage_level) should be from oemof 0.3.0. 

We have the issue to upgrade the oemof version, but we should do that in an orderly way:  
* Check that only changing the version and respective parameter names results in the same simulation output.
* Update requirements.txt

I would suggest that you revert to the old oemof version for this pull request. Please note down these changes. Ideally, you would store them in an own branch: "feature_update_oemof_version".

File: G2a_oemof_busses_and_componets.py [hunk 5]
```diff
@@ -361,9 +361,9 @@ def storage_oem(micro_grid_system, bus_electricity_dc, experiment):
                 variable_costs=experiment['storage_cost_var'])},
             outputs                         = {bus_electricity_dc: solph.Flow(
                 investment=solph.Investment(ep_costs=experiment['storage_power_cost_annuity']))},
-            capacity_loss                   = experiment['storage_loss_timestep'],  # from timestep to timestep
-            capacity_min                    = experiment['storage_soc_min'],
-            capacity_max                    = experiment['storage_soc_max'],
+            loss_rate                   = experiment['storage_loss_timestep'],  # from timestep to timestep
+            min_storage_level                    = experiment['storage_soc_min'],
+            max_storage_level                    = experiment['storage_soc_max'],
```
Comment (smartie2076) [inline G2a_oemof_busses_and_componets.py]: see above.

File: G2b_constraints_custom.py [hunk 6]
```diff
@@ -174,6 +174,11 @@ def stability_rule_capacity(model, t):
                 expr += stored_electricity * experiment['storage_Crate_discharge'] \
                         * experiment['storage_efficiency_discharge'] \
                         * experiment['inverter_dc_ac_efficiency']
+
+            ## ------- Get pv t ------- #
+            if case_dict['pv_fixed_capacity'] != None:
+                expr += model.flow[pv, el_bus_dc, t] * experiment['inverter_dc_ac_efficiency']
```
Comment (smartie2076) [inline G2b_constraints_custom.py]: Why do you add this to the stability constraint?
Stability is only induced (in this criterion) by (a) active diesel generation (b) potential storage discharge. PV with its variability is from my perspective not able to stabilize the system (or ensure its robustness to pv variability, in fact...).

File: G2b_constraints_custom.py [hunk 7]
```diff
@@ -207,6 +212,12 @@ def stability_rule_power(model, t):
                     logging.warning("Error: 'storage_fixed_power' can only be None, False or float.")
 
                 expr += storage_power * experiment['inverter_dc_ac_efficiency']
+
+            ## ------- Get pv t ------- #
+            if case_dict['pv_fixed_capacity'] != None:
+                expr += model.flow[pv, el_bus_dc, t] * experiment['inverter_dc_ac_efficiency']
+
```
Comment (smartie2076) [inline G2b_constraints_custom.py]: See comment above. Please explain to be in our live session why you add this to the hybrid constraint.

File: G2b_constraints_custom.py [hunk 8]
```diff
@@ -635,6 +646,9 @@ def stability_per_timestep_rule(model, t):
 
             return (expr >= 0)
 
-        model.stability_constraint = po.Constraint(model.TIMESTEPS, rule=stability_per_timestep_rule)
+        if bus_type == 'ac':
+            model.stability_constraint_timestep_ac = po.Constraint(model.TIMESTEPS, rule=stability_per_timestep_rule)
+        else:
+            model.stability_constraint_timestep_dc = po.Constraint(model.TIMESTEPS, rule=stability_per_timestep_rule)
```
Comment (smartie2076) [inline G2b_constraints_custom.py]: Are you limiting the stability constraint here and saying it per definition is only applicable to the AC bus?
Did it create two constraints somewhere or is this a basic simulation decision you would make?

File: H0_multicriteria_analysis.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
+'''
+1. Agafar resultats de costos i capacitats
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: Please change to english

File: H0_multicriteria_analysis.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+import xlsxwriter
+import os
+
+class Multicriteria():
+
+    def main_analysis(all_results,project_locations,shortage_levels):
+        # keys of dictionary all_results are the filenames, and values are the dictionaries with the results of every case and experiment
+
+        # qualitative indicators
+        T2_punctuations = {'pv': 4, 'wind': 0, 'maingrid': 4, 'genset': 2.5}
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: I think I would prefer to have this in the input template, so that the Offgridders user can change this his/herself. We could have it at default-values though.
Possible todos:
* Create tab in input template, choices: Scalar value, or "default"
* Code access to those punctuations if "default" used for specific components

Comment (smartie2076) [reply H0_multicriteria_analysis.py]: Likewise for definition of punctuations below.
Could accept it as a preliminaty H_xx feature, but in the end I would like variable input better.

File: H0_multicriteria_analysis.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+        Compromise.ranking_method(evaluations,project_locations,shortage_levels,cases)
+
+        return
+
+
+class Compromise():
+    def ranking_method(evaluations,project_locations,shortage_levels,cases):
+
+        # Dimensions and criteria weights (have to be adjusted)
+        weights_dimensions = {'economic':0.2,'technical':0.2,'social':0.2,'environmental':0.2,'institutional':0.2}
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: weights should also be editable via the input file (in the end)
-> if we do not address this now, we should hint at this need to improve in an issue.

File: H0_multicriteria_analysis.py [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        return {'economic':economic_eval,'technical':technical_eval,'social':social_eval,
+                                  'environmental':environmental_eval,'institutional':institutional_eval}
+
+
+    def matrix_results_evaluation(all_global_evaluations,project_locations,shortage_levels,cases,Ls,L1s,Linfs):
+
+        columns = {1: 'A', 2: 'B', 3: 'C', 4: 'D', 5: 'E', 6: 'F', 7: 'G', 8: 'H', 9: 'I', 10: 'J'
+            , 11: 'K', 12: 'L', 13: 'M', 14: 'N', 15: 'O', 16: 'P', 17: 'Q', 18: 'R', 19: 'S', 20: 'T'
+            , 21: 'U', 22: 'V', 23: 'W', 24: 'X', 25: 'Y', 26: 'Z'}
+
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: What is this for?

File: H0_multicriteria_analysis.py [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+        if not (os.path.isdir('./MCDA_evaluation_results')):
+            os.mkdir('./MCDA_evaluation_results')
+
+        workbook = xlsxwriter.Workbook('./MCDA_evaluation_results/Evaluation_results.xlsx')
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: This stores in the main folder level, correct?
The MCDA results should be included in the simulation results/test/ folder. The "test" of this is user-defined. This path should therefore be related to the user input path (check experiment["output_folder"])!

File: H0_multicriteria_analysis.py [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+                worksheet.write(row, col, element, merge_format)
+                col += 1
+
+        workbook.close()
+
+        return
+
+
+    def presentation(all_results,shortage_levels,oemof_results,experiment):
+
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: I do not understand why this function is called "presentation"?

File: A0_main_script.py [hunk 15]
```diff
@@ -90,10 +90,15 @@
...
 from F_case_definitions import cases
 from G0_oemof_simulate import oemof_simulate
+from H0_multicriteria_analysis import multicriteria_helpers
 
 experiment_count = 0
 total_number_of_simulations = settings['total_number_of_experiments'] * len(case_list)
 
+# variables necessary to store results for subsequent multicriteria analysis
+all_results = {}
+shortage_levels = []
```
Comment (smartie2076) [inline A0_main_script.py]: Why is this not in overall_results, but has an additional storage?

File: A0_main_script.py [hunk 16]
```diff
@@ -130,6 +136,10 @@
         # Run simulation, evaluate results
         oemof_results = oemof_simulate.run(sensitivity_experiment_s[experiment], experiment_case_dict)
 
+        # Store results for subsequent multicriteria analysis
+        multicriteria_helpers.presentation(all_results,shortage_levels,oemof_results,sensitivity_experiment_s[experiment])
```
Comment (smartie2076) [inline A0_main_script.py]: If all_results could be gathers in overall_results, this function could be integrated into 
    'overall_results = helpers.store_result_matrix(overall_results, sensitivity_experiment_s[experiment], oemof_results)'

If function stays: Rename from presentation into update_matrix_all_results or something else that describes the function better

File: F_case_definitions.py [hunk 17]
```diff
@@ -124,6 +124,16 @@ def update_dict(capacities_oem, specific_case, experiment):
...
 
+        ###########################################
+        # Include shortage timestep constraint    #
+        ###########################################
+
+        if specific_case['shortage_timestep_constraint'] == True:
+            experiment_case_dict.update({'shortage_timestep_constraint':True})
+        else:
+            experiment_case_dict.update({'shortage_timestep_constraint': False})
+
```
Comment (smartie2076) [inline F_case_definitions.py]: ok

File: G1_oemof_create_model.py [hunk 18]
```diff
@@ -295,18 +296,21 @@ def build(experiment, case_dict):
...
             if bus_electricity_dc != None:
                 shortage_constraints.timestep(model, case_dict, experiment, sink_demand_dc, 
-                                              source_shortage, bus_electricity_dc)
-        '''
+                                              source_shortage, bus_electricity_dc,'dc')
+
+
         return micro_grid_system, model
 
+
```
Comment (smartie2076) [inline G1_oemof_create_model.py]: code ok, but is is necessary to have this shortage constraint for both busses at the same time?
What would be the implications?

File: G3a_economic_evaluation.py [hunk 19]
```diff
@@ -183,6 +222,9 @@ def revenue_main_grid_feedin(oemof_results, experiment):
         oemof_results.update({'revenue_main_grid_feedin_annual':
                 - oemof_results['feedin_main_grid_mg_side_annual_kWh'] * experiment['maingrid_feedin_tariff']})
 
+        oemof_results.update({'operation_mantainance_expenditures': oemof_results['operation_mantainance_expenditures']
+                                                                    + oemof_results['revenue_main_grid_feedin_annual']})
+
```
Comment (smartie2076) [inline G3a_economic_evaluation.py]: For me, "revenue" should not be substracted from expenditures but be an own value.

File: G3a_economic_evaluation.py [hunk 20]
```diff
@@ -78,29 +78,59 @@ def annuities_365(case_dict, oemof_results, experiment):
...
+                          'wind',
+                          'genset',
+                          'storage',
+                          'pcoupling',
+                          'maingrid_extension',
+                          'distribution_grid',
+                          'rectifier_ac_dc',
+                          'inverter_dc_ac',
+                          'project']
+
```
Comment (smartie2076) [inline G3a_economic_evaluation.py]: below righthand site, there are a lot of new lines, while none seem to be deleted on the left side. However, isn't this info already calculated somewhere? Sorry, hard to read in this editor on github. also maybe content of next commits...

File: A0_main_script.py (hunk 1 above)
Comment (marc-juanpera) [reply A0_main_script.py]: The input file does not change very much. Only one more tab will be added to set the parameters for the multicriteria analysis (weights and qualitative punctuations) and some more lines are added in the case definition tab.

File: G2a_oemof_busses_and_componets.py (hunk 4 above)
Comment (marc-juanpera) [reply G2a_oemof_busses_and_componets.py]: yes, I needed 0.3 version and I forgot to change it back when pushing the files. This is one of the reasons I think it is better to work in paralel

File: G2b_constraints_custom.py (hunk 6 above)
Comment (marc-juanpera) [reply G2b_constraints_custom.py]: This change sould not be accepted.
I included this when I analysed the communities of Peru. There, PV source is actually more robust than diesel because there are significant lacks of fuel supply. I should have removed when analysing again Nigeria communities.

File: H0_multicriteria_analysis.py (hunk 10 above)
Comment (marc-juanpera) [reply H0_multicriteria_analysis.py]: I want to discuss the format of the excel tab dedicated to inputs of multicriteria analysis on Tuesday.

File: H0_multicriteria_analysis.py (hunk 12 above)
Comment (marc-juanpera) [reply H0_multicriteria_analysis.py]: I needed that for a loop below to select the excel cells. I didn't know a nicer way.

File: H0_multicriteria_analysis.py (hunk 13 above)
Comment (marc-juanpera) [reply H0_multicriteria_analysis.py]: In the version I am working on it is already like you say.

File: H0_multicriteria_analysis.py (hunk 14 above)
Comment (marc-juanpera) [reply H0_multicriteria_analysis.py]: Agree, maybe a more accurate name would be multicriteria_input_preparation

File: A0_main_script.py (hunk 15 above)
Comment (marc-juanpera) [reply A0_main_script.py]: This was the nicest way I found to get the results I needed. But when we structure everything, I want to clean the code and this should be solved. That is another reason for not merging anything now.

File: A0_main_script.py (hunk 16 above)
Comment (marc-juanpera) [reply A0_main_script.py]: I will look at it

File: G3a_economic_evaluation.py (hunk 19 above)
Comment (marc-juanpera) [reply G3a_economic_evaluation.py]: if it is separate value, it would not be included in the multicriteria evaluation. At least for MCDA it has to be added.

File: A0_main_script.py [hunk 21]
```diff
@@ -137,7 +138,8 @@
         oemof_results = oemof_simulate.run(sensitivity_experiment_s[experiment], experiment_case_dict)
 
         # Store results for subsequent multicriteria analysis
-        multicriteria_helpers.presentation(all_results,shortage_levels,oemof_results,sensitivity_experiment_s[experiment])
+        if settings['perform_multicriteria_analysis'] == True:
+            multicriteria_helpers.presentation(all_results,shortage_levels,oemof_results,sensitivity_experiment_s[experiment])
```
Comment (smartie2076) [inline A0_main_script.py]: Nice option. Included in the xlsx, I presume.

File: B_read_from_files.py [hunk 22]
```diff
@@ -131,6 +131,8 @@ def get_case_definitions(file, sheet_project_sites):
                 case_definitions[case][key] = excel_template.identify_true_false(case_definitions[case][key])
             if case_definitions[case]['max_shortage'] != 'default':
                 case_definitions[case].update({'max_shortage': float(case_definitions[case]['max_shortage'])})
+            if case_definitions[case]['shortage_penalty_costs'] != 'default':
+                case_definitions[case].update({'shortage_penalty_costs': float(case_definitions[case]['shortage_penalty_costs'])})
```
Comment (smartie2076) [inline B_read_from_files.py]: is this possible per case definition now?

File: G2a_oemof_busses_and_componets.py [hunk 23]
```diff
@@ -38,11 +38,11 @@ def shortage(micro_grid_system, bus_electricity_ac, bus_electricity_dc, experime
         source_shortage = solph.Source(label="source_shortage",
                                        outputs={
                                            bus_electricity_ac: solph.Flow(
-                                               variable_costs   = experiment['shortage_penalty_costs'],
+                                               variable_costs   = case_dict['shortage_penalty_costs'],
                                                nominal_value    = case_dict['max_shortage'] * case_dict['total_demand_ac'],
                                                summed_max       = 1),
                                            bus_electricity_dc: solph.Flow(
-                                               variable_costs   = experiment['shortage_penalty_costs'],
+                                               variable_costs   = case_dict['shortage_penalty_costs'],
```
Comment (smartie2076) [inline G2a_oemof_busses_and_componets.py]: Is it still possible to perform a sensitivity analysis on the shortage penalty costs? 
Why is this necessary?

File: F_case_definitions.py [hunk 24]
```diff
@@ -102,17 +102,22 @@ def update_dict(capacities_oem, specific_case, experiment):
...
+                experiment_case_dict.update({'max_shortage': experiment['shortage_max_allowed']})
+            elif (isinstance(specific_case['max_shortage'], float) or isinstance(specific_case['max_shortage'], int)):
                 experiment_case_dict.update({'allow_shortage': True})
                 experiment_case_dict.update({'max_shortage': specific_case['max_shortage']})
+            if specific_case['shortage_penalty_costs']== 'default':
+                experiment_case_dict.update({'allow_shortage': True})
+                experiment_case_dict.update({'shortage_penalty_costs': experiment['shortage_penalty_costs']})
+            elif (isinstance(specific_case['shortage_penalty_costs'], float) or isinstance(specific_case['shortage_penalty_costs'], int)):
+                experiment_case_dict.update({'allow_shortage': True})
+                experiment_case_dict.update({'shortage_penalty_costs': specific_case['shortage_penalty_costs']})
```
Comment (smartie2076) [inline F_case_definitions.py]: why is this necessary?

File: H0_multicriteria_analysis.py [hunk 25]
```diff
@@ -37,22 +43,31 @@ def main_analysis(all_results,project_locations,shortage_levels):
...
+        nonreliable_capacities_projects = []
 
         for project_site in all_results:
             project_evaluations = {}
-            cases = []
+            cases = {'reliable':[],'nonreliable':[]}
+            reliable_capacities = {'capacity_pv_kWp':[],'capacity_wind_kW':[],
+                                   'capacity_storage_kWh':[],'capacity_genset_kW':[],'capacity_pcoupling_kW':[]}
+            nonreliable_capacities = {'capacity_pv_kWp': [], 'capacity_wind_kW': [],
+                                   'capacity_storage_kWh': [], 'capacity_genset_kW': [], 'capacity_pcoupling_kW': []}
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: what does "reliable" and "nonreliable" capacities mean?

Comment (smartie2076) [reply H0_multicriteria_analysis.py]: Is it connected to shortage levels, ie. "reliable" capacities are stored from a 100% reliable system?

File: H0_multicriteria_analysis.py [hunk 26]
```diff
@@ -260,92 +391,315 @@ def create_diccionary(self):
...
+            , 10: 'K', 11: 'L', 12: 'M', 13: 'N', 14: 'O', 15: 'P', 16: 'Q', 17: 'R', 18: 'S', 19: 'T'
+            , 20: 'U', 21: 'V', 22: 'W', 23: 'X', 24: 'Y', 25: 'Z'}
 
-        workbook = xlsxwriter.Workbook('./MCDA_evaluation_results/Evaluation_results.xlsx')
+
+        #if not (os.path.isdir('./MCDA_evaluation_results')):
+        #    os.mkdir('./MCDA_evaluation_results')
+
+
+        workbook = xlsxwriter.Workbook(settings['output_folder']+'/MCDA_evaluation_results.xlsx')
```
Comment (smartie2076) [inline H0_multicriteria_analysis.py]: Ah, okay, here you now include the user-defined folder.

Comment (smartie2076) [reply H0_multicriteria_analysis.py]: I would separate the functions that write to the workbook into H1_output, so that the equations in the MCDA are separated from the actual writing/formatting of the xsls file. Else, one does not find the actual important equations.

File: A0_main_script.py (hunk 1 above)
Comment (smartie2076) [reply A0_main_script.py]: Sounds good. Looking forward to look at the xlsx

**Comment by smartie2076** [comment]
Current status: Pull not to be merged into dev. Face-to-Face discussion follows.

**Comment by smartie2076** [comment]
Marc, can  you create an issue in which you describe what you are planning to integrade into Offgridders? (Just copy&paste one of the work plans ie. that you have created already) We would close the issue later with merging this pull request.

**Comment by smartie2076** [comment]
Closing PR as discussed live in skype
